File: src/ObjectManager/Object.Tes/FilePacks/Records/300-REPA.Repair item.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class REPARecord : Record, IHaveEDID, IHaveMODL
    {
        public struct RIDTField
        {
            public float Weight;
            public int Value;
            public int Uses;
            public float Quality;

            public RIDTField(UnityBinaryReader r, int dataSize)
            {
                Weight = r.ReadLESingle();
                Value = r.ReadLEInt32();
                Uses = r.ReadLEInt32();
                Quality = r.ReadLESingle();
            }
        }

        public override string ToString() => $"REPA: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model Name
        public STRVField FNAM; // Item Name
        public RIDTField RIDT; // Repair Data
        public FILEField ICON; // Inventory Icon
        public FMIDField<SCPTRecord> SCRI; // Script Name

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            if (format == GameFormatId.TES3)
                switch (type)
                {
                    case "NAME": EDID = new STRVField(r, dataSize); return true;
                    case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                    case "FNAM": FNAM = new STRVField(r, dataSize); return true;
                    case "RIDT": RIDT = new RIDTField(r, dataSize); return true;
                    case "ITEX": ICON = new FILEField(r, dataSize); return true;
                    case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                    default: return false;
                }
            return false;
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/300-REPA.Repair item.swift
//
//  REPARecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class REPARecord: Record, IHaveEDID, IHaveMODL {
    public struct RIDTField {
        public let weight: Float
        public let value: Int32
        public let uses: Int32
        public let quality: Float

        init(_ r: BinaryReader, _ dataSize: Int) {
            weight = r.readLESingle()
            value = r.readLEInt32()
            uses = r.readLEInt32()
            quality = r.readLESingle()
        }
    }

    public override var description: String { return "REPA: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model Name
    public var FNAM: STRVField! // Item Name
    public var RIDT: RIDTField! // Repair Data
    public var ICON: FILEField! // Inventory Icon
    public var SCRI: FMIDField<SCPTRecord>! // Script Name

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        guard format == .TES3 else {
            return false
        }
        switch type {
        case "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "FNAM": FNAM = r.readSTRV(dataSize)
        case "RIDT": RIDT = RIDTField(r, dataSize)
        case "ITEX": ICON = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        default: return false
        }
        return true
    }
}
